backend/push.js

```diff
@@ -1,53 +1,292 @@
+        })
+      }
+    }
+    socket.pushSubscriptionId = subscriptionId
+    server.pushSubscriptions[subscriptionId].subscriptions.forEach((channelID) => {
+      server.subscribersByChannelID?.[channelID].delete(server.pushSubscriptions[subscriptionId])
+    })
+    server.pushSubscriptions[subscriptionId].sockets.add(socket)
+    socket.subscriptions?.forEach(channelID => {
+      server.pushSubscriptions[subscriptionId].subscriptions.add(channelID)
+    })
+    saveSubscription(server, subscriptionId)
   },
-  [PUSH_SERVER_ACTION_TYPE.DELETE_SUBSCRIPTION] (payload) {
+  [PUSH_SERVER_ACTION_TYPE.DELETE_SUBSCRIPTION] () {
     const socket = this
-    const subscriptionId = JSON.parse(payload)
+    const { server, pushSubscriptionId: subscriptionId } = socket
 
-    delete socket.server.pushSubscriptions[subscriptionId]
-  },
-  [PUSH_SERVER_ACTION_TYPE.SEND_PUSH_NOTIFICATION]: async function (payload) {
-    const pushSubscriptions = this.server.pushSubscriptions
-    const data = JSON.parse(payload)
-    const sendPush = (sub) => pushController.sendNotification(
-      sub, JSON.stringify({ title: data.title, body: data.body })
-    )
+    if (subscriptionId) {
+      removeSubscription(server, subscriptionId)
+    }
+  }
+}
 
-    // NOTE: if the payload contains 'endpoint' field, send push-notification to that particular subscription.
-    //       otherwise, iterate all existing subscriptions and broadcast the push-notification to all.
+const removeSubscription = (server, subscriptionId) => {
+  const subscription = server.pushSubscriptions[subscriptionId]
+  delete server.pushSubscriptions[subscriptionId]
+  if (server.subscribersByChannelID) {
+    subscription.subscriptions.forEach((channelID) => {
+      server.subscribersByChannelID[channelID].delete(subscription)
+    })
+  }
+  deleteSubscriptionFromIndex(subscriptionId).then(() => {
+    return sbp('chelonia/db/delete', `_private_webpush_${subscriptionId}`)
+  }).catch((e) => console.error(e, 'Error removing subscription', subscriptionId))
+}
 
-    if (data.endpoint) {
-      const subscription = pushSubscriptions[data.endpoint]
+const deleteClient = (subscriptionId) => {
+  const server = sbp('okTurtles.data/get', PUBSUB_INSTANCE)
+  removeSubscription(server, subscriptionId)
+}
 
-      await sendPush(subscription)
-    } else {
-      for (const subscription of Object.values(pushSubscriptions)) {
-        await sendPush(subscription)
-      }
+const encryptPayload = async (subcription: Object, data: string) => {
+  const readableStream = new Response(data).body
+  const [asPublic, IKM] = await subcription.encryptionKeys
+
+  return encrypt(aes128gcm, readableStream, 32768, asPublic, IKM).then(async (bodyStream) => {
+    const chunks = []
+    const reader = bodyStream.getReader()
+    for (;;) {
+      const { done, value } = await reader.read()
+      if (done) break
+      chunks.push(new Uint8Array(value))
+    }
+    return Buffer.concat(chunks)
+  })
+}
```

Comment:
Above this function definition, could you please add a comment explaining what keys are used to encrypted this payload and what the trust assumptions are here?

I.e. my understanding is that the keys that are used are keys that were generated by the browser (i.e. untrusted code we have no control over) that were sent over TLS to the server over the websocket. Could a TLD MITM have replaced them? If so, please mention that in the comment too.

Finally, please add a sentence or two explaining what information is exposed if these key were compromised.